power_grid_model_c/power_grid_model/include/power_grid_model/job_dispatch_interface.hpp

```diff
@@ -0,0 +1,54 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+// batch dispatch interface class
+
+#include "common/calculation_info.hpp"
+#include "common/common.hpp"
+
+#include <concepts>
+#include <type_traits>
+#include <utility>
+
+namespace power_grid_model {
+template <typename Adapter> class JobDispatchInterface {
+  public:
+    template <typename Calculate, typename ResultDataset>
+        requires requires(Adapter& adapter, Calculate&& calculation_fn, ResultDataset const& result_data, Idx pos) {
+            { adapter.calculate_impl(std::forward<Calculate>(calculation_fn), result_data, pos) } -> std::same_as<void>;
+        }
+    void calculate(Calculate&& calculation_fn, ResultDataset const& result_data, Idx pos = 0) {
+        return static_cast<Adapter*>(this)->calculate_impl(std::forward<Calculate>(calculation_fn), result_data, pos);
+    }
+
+    template <typename Calculate>
+        requires requires(Adapter& adapter, Calculate&& calculation_fn) {
+            { adapter.cache_calculate_impl(std::forward<Calculate>(calculation_fn)) } -> std::same_as<void>;
+        }
+    void cache_calculate(Calculate&& calculation_fn) {
+        return static_cast<Adapter*>(this)->cache_calculate_impl(std::forward<Calculate>(calculation_fn));
+    }
+
+    template <typename UpdateDataset> void prepare_job_dispatch(UpdateDataset const& update_data) {
+        return static_cast<Adapter*>(this)->prepare_job_dispatch_impl(update_data);
+    }
+
+    template <typename UpdateDataset> void setup(UpdateDataset const& update_data, Idx scenario_idx) {
+        return static_cast<Adapter*>(this)->setup_impl(update_data, scenario_idx);
+    }
+
+    void winddown() { return static_cast<Adapter*>(this)->winddown_impl(); }
+
+    CalculationInfo get_calculation_info() const {
+        return static_cast<const Adapter*>(this)->get_calculation_info_impl();
+    }
+
+    void thread_safe_add_calculation_info(CalculationInfo const& info) {
+        static_cast<Adapter*>(this)->thread_safe_add_calculation_info_impl(info);
+    }
+};
```

Comment:
make sure it's not accidentally constructible or destructible in a way that leaks memory

```suggestion

  private:
    JobDispatchInterface() = default;
    JobDispatchInterface(const JobDispatchInterface& /*other*/) = default;
    JobDispatchInterface(JobDispatchInterface&& /*other*/) = default;
    JobDispatchInterface& operator=(const JobDispatchInterface& /*other*/) = default;
    JobDispatchInterface& operator=(JobDispatchInterface&& /*other*/) = default;
    ~JobDispatchInterface() = default;

    friend Adapter;
};
```

Reply:
We were just talking about this 😆 . Will do!

Reply:
please fix the new sonar cloud issues: https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=1036&id=PowerGridModel_power-grid-model&open=AZiIyo1L1VB51fmTBV-X

Reply:
@nitbharambe Will address them in his changes :)